### PR TITLE
feat(icon): support for iOS splash screens

### PR DESCRIPTION
**Pull request by dvarnai**
Hi,

This pull request adds support to automatically generate splash screens for IOS devices (#101 ) as the current implementation doesn't work as it's ignored by IOS.

The changes might not be structured in the nicest way so feel free to let me know what you think and I'll refactor it.

Best regards,
Daniel

### CODE REVIEW

**Comment by RubenVanEldik** [comment]
I was just looking for this! @dvarnai Awesome that you made a pull request. Hopefully it will be merged soon. :)

File: lib/icon/module.js [hunk 1]
```diff
@@ -19,6 +19,18 @@ async function run (pwa, _emitAssets) {
...
+      ['ipadpro1', 1536, 2048], // Ipad Pro 9.7"
+      ['ipadpro2', 1668, 2224], // Ipad Pro 10.5"
+      ['ipadpro3', 2048, 2732], // Ipad Pro 12.9"
+      ['iphonese', 640, 1136], // Iphone SE
+      ['iphone6', 750, 1334], // Iphone 6
+      ['iphoneplus', 1080, 1920], // Iphone Plus
+      ['iphonex', 1125, 2436], // Iphone X
+      ['iphonexr', 828, 1792], // Iphone XR
+      ['iphonexsmax', 1242, 2688] // Iphone XS Max
+    ],
```
Comment (RubenVanEldik) [inline lib/icon/module.js:33]:
```suggestion
    iosSizes: [
      { name: 'ipad', width: 1536, height: 2048, pixelRatio: 2 },
      { name: 'ipad', width: 1536, height: 2048, pixelRatio: 2 }, // 9.7" iPad
      { name: 'ipad2', width: 1620, height: 2160, pixelRatio: 2 }, // 10.2" iPad
      { name: 'ipadpro1', width: 1536, height: 2048, pixelRatio: 2 }, // 9.7" iPad Air and 9.7" iPad Pro
      { name: 'ipadpro2', width: 1668, height: 2224, pixelRatio: 2 }, // 10.5" and 11" iPad Pro and 10.5" iPad Air
      { name: 'ipadpro3', width: 2048, height: 2732, pixelRatio: 2 }, // 12.9" iPad Pro
      { name: 'iphonese', width: 640, height: 1136, pixelRatio: 2 }, // 4" iPhone SE
      { name: 'iphone6', width: 750, height: 1334, pixelRatio: 2 }, // iPhone 6, 6S, 7, and 8, and 4.7" SE
      { name: 'iphoneplus', width: 1080, height: 1920, pixelRatio: 2 }, // iPhone 6 Plus, 6s Plus, 7 Plus, and 8 Plus
      { name: 'iphonex', width: 1125, height: 2436, pixelRatio: 2 }, // iPhone X, XS, and 11 Pro
      { name: 'iphonexr', width: 828, height: 1792, pixelRatio: 2 }, // iPhone XR and 11
      { name: 'iphonexsmax', width: 1242, height: 2688, pixelRatio: 2 } // iPhone XS Max and 11 Pro Max
    ],
```

File: lib/icon/module.js [hunk 2]
```diff
@@ -134,15 +150,41 @@ async function generateIcons (options) {
...
+    options.iconHash = await hasha.fromFile(options.iconSrc).then(h => h.substring(0, 6))
+  }
+  if (!options._cacheDir) {
+    options._cacheDir = path.join(__dirname, '.cache', options.iconHash)
+  }
+
+  // Generate _iosSplash
+  options._iosSplash = {}
+  for (const size of options.iosSizes) {
+    options._iosSplash[size[0]] = `${options.targetDir}/icon_${size[0]}.${options.iconHash}.png`
```
Comment (RubenVanEldik) [inline lib/icon/module.js:165]:
```suggestion
  for (const { name } of options.iosSizes) {
    options._iosSplash[name] = `${options.targetDir}/icon_${name}.${options.iconHash}.png`
```

File: lib/icon/module.js [hunk 3]
```diff
@@ -134,15 +150,41 @@ async function generateIcons (options) {
...
+  }
+
+  // Generate _assetIosSplash
+  options._assetIosSplash = {}
+  options.iosSizes.forEach((size) => {
+    options._assetIosSplash[size[0]] = joinUrl(options.publicPath, options._iosSplash[size[0]])
+  })
```
Comment (RubenVanEldik) [inline lib/icon/module.js:172]:
```suggestion
  options._assetIosSplash = options.iosSize.map(size => ({
    ...size,
    href: joinUrl(options.publicPath, options._iosSplash[name])
  }))
```

File: lib/icon/module.js [hunk 4]
```diff
@@ -156,12 +198,22 @@ function emitAssets (options) {
...
               const targetFilename = options._icons[size]
               const srcFileName = path.join(options._cacheDir, `${size}.png`)
               const src = await fs.readFile(srcFileName)
               compilation.assets[targetFilename] = { source: () => src, size: () => src.length }
             }))
+
+            // Ios splash
+            await Promise.all(options.iosSizes.map(async (size) => {
+              const targetFilename = options._iosSplash[size[0]]
+              const srcFileName = path.join(options._cacheDir, `${size[0]}.png`)
```
Comment (RubenVanEldik) [inline lib/icon/module.js:213]:
```suggestion
            await Promise.all(options.iosSizes.map(async ({ name }) => {
              const targetFilename = options._iosSplash[name]
              const srcFileName = path.join(options._cacheDir, `${name}.png`)
```

File: lib/icon/resize.js [hunk 5]
```diff
@@ -1,15 +1,24 @@
...
   await Promise.all(sizes.map((size) => {
     const distFile = path.join(distDir, `${size}.png`)
     return new Promise((resolve) => {
       inputFile.clone().contain(size, size).write(distFile, () => resolve())
     })
   }))
+
+  // IOS Splash Screens
+  await Promise.all(iosSizes.map((iosSize) => {
+    const distFile = path.join(distDir, `${iosSize[0]}.png`)
```
Comment (RubenVanEldik) [inline lib/icon/resize.js:17]:
```suggestion
  await Promise.all(iosSizes.map(({ name, width, height }) => {
    const distFile = path.join(distDir, `${name}.png`)
```

File: lib/icon/resize.js [hunk 6]
```diff
@@ -1,15 +1,24 @@
...
+    return new Promise((resolve) => {
+      inputFile.clone().contain(iosSize[1], iosSize[2]).write(distFile, () => resolve())
```
Comment (RubenVanEldik) [inline lib/icon/resize.js:19]:
```suggestion
      inputFile.clone().contain(width, height).write(distFile, () => resolve())
```

File: lib/meta/module.js [hunk 7]
```diff
@@ -94,7 +94,16 @@ function generateMeta (pwa) {
...
+      this.options.head.link.push({ href: options.iosSplash.iphonexsmax, media: '(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3)', rel: 'apple-touch-startup-image' })
+      this.options.head.link.push({ href: options.iosSplash.ipad, media: '(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2)', rel: 'apple-touch-startup-image' })
+      this.options.head.link.push({ href: options.iosSplash.ipadpro1, media: '(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2)', rel: 'apple-touch-startup-image' })
+      this.options.head.link.push({ href: options.iosSplash.ipadpro2, media: '(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2)', rel: 'apple-touch-startup-image' })
+      this.options.head.link.push({ href: options.iosSplash.ipadpro3, media: '(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2)', rel: 'apple-touch-startup-image' })
```
Comment (RubenVanEldik) [inline lib/meta/module.js:106]:
```suggestion
      for (const { href, width, height, pixelRatio } of options.iosSplash) {
        this.options.head.linnk.push({
          href,
          media: `(device-width: ${Math.round(width / pixelRatio)}px)
            and (device-height: ${Math.round(height / pixelRatio)}px)
            and (-webkit-device-pixel-ratio: ${pixelRatio})`,
          rel: 'apple-touch-startup-image'
        })
      }
```

Comment (dvarnai) [reply lib/meta/module.js:106]: Great idea, I didn't realise there was a pixel ratio used for those dimensions!

**Comment by dvarnai** [comment]
@RubenVanEldik cheers! great idea!

**Comment by pi0** [comment]
@dvarnai I've made some refactors. Would be also nice if later on you could help on adding some docs.

**Comment by shou-kato** [comment]
Is this feature available now?

**Comment by shou-kato** [comment]
Is this feature available now?

File: lib/icon/module.js [hunk 8]
```diff
@@ -19,6 +19,18 @@ async function run (pwa, _emitAssets) {
   const defaults = {
     iconFileName: 'icon.png',
     sizes: [64, 120, 144, 152, 192, 384, 512],
+    iosSizes: [
+      [1536, 2048, 'ipad'], // Ipad
+      [1536, 2048, 'ipadpro9'], // Ipad Pro 9.7"
+      [1668, 2224, 'ipadpro10'], // Ipad Pro 10.5"
+      [2048, 2732, 'ipadpro12'], // Ipad Pro 12.9"
+      [640, 1136, 'iphonese'], // Iphone SE
+      [50, 1334, 'iphone6'], // Iphone 6
```
Comment (adambullmer) [inline lib/icon/module.js:28]: Is the iphone 6 splash screen suppose to be 50 px wide?

Comment (hanzoz) [reply lib/icon/module.js:28]: definitely is 750 x 1334

**Pull request merged**